In [1]:
#!pip install tensorflow_hub
#!pip install tensorflow_text

In [2]:
import tensorflow_hub as hub
import tensorflow_text as text         # Registers the ops.

- for each of this encoder model url thier is corresponding preprocessor url, which the link is given in the end of the this notebook

In [3]:
preprocess_url = "https://kaggle.com/models/tensorflow/bert/frameworks/TensorFlow2/variations/en-uncased-preprocess/versions/3"
encoder_url = "https://www.kaggle.com/models/tensorflow/bert/frameworks/TensorFlow2/variations/en-uncased-l-12-h-768-a-12/versions/4"

In [4]:
# Create hub layer by passing the url

bert_preprocess_model = hub.KerasLayer(preprocess_url)
bert_encoder = hub.KerasLayer(encoder_url)

In [5]:
# let's apply some statements and do preprocessing on those statements

test_text = ['nice movie', 'i love reading books']   # Create (word or sentence)embedding for this statement 

text_preprocess = bert_preprocess_model(test_text) 
text_preprocess.keys()                                            # Text_Preprocess is gona be in dictionary

dict_keys(['input_word_ids', 'input_type_ids', 'input_mask'])

_It preproceed those 2 sentence and produced this 3 objects
lets see the individual element in the dictionary_

In [6]:
text_preprocess['input_mask']

# "input_mask": has value 1 at the position of all input tokens present before padding and value 0 for the padding tokens.

<tf.Tensor: shape=(2, 128), dtype=int32, numpy=
array([[1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]])>

1. The encoder_inputs are a dict of three int32 Tensors, all with shape [batch_size, seq_length], whose elements represent the batch of input sequences as follows:  
  
2. The shape is (2,128)[batch_size, seq_length] batchsize = 2 & seq_length = 128, why 2 because we have 2-sentence and 128 is  maximum length of the sentence, you can see it as 2 mask for 2 sentence  
  
3. The first sentence is 2 word but the mask as 4 word [1, 1, 1, 1], lets try to understand that.. 
    - the way the word works is it will always put a special token [CLS] in the beginning and to seperate 2 sentences it will put a special token [SEP]   
  
>" [CLS] to indicate the beginning of the sentence, [SEP] to separate multiple sentences, and [PAD] to make each sentence have the same number of tokens "  
  
So now if you count tokens   
1. ==> [ CLS nice movie SEP ] ==> [ 1, 1, 1, 1 ] - (4_ones)
2. ==> [ CLS i love reading books SEP ] ==> [ 1, 1, 1, 1, 1, 1 ] - (6_ones) 

In [7]:
text_preprocess['input_type_ids']

# "input_type_ids": has the index of the input segment that gave rise to the input token at the respective position.
# The first input segment (index 0) includes the start-of-sequence token and its end-of-segment token. 
# The second segment (index 1, if present) includes its end-of-segment token. Padding tokens get index 0 again.

<tf.Tensor: shape=(2, 128), dtype=int32, numpy=
array([[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]])>

- input_type_ids are very useful for multiple sentenceses in 1 statement for ourcase it's not intresting everything will be in 0

In [8]:
text_preprocess['input_word_ids']

# "input_word_ids": has the token ids of the input sequences.

<tf.Tensor: shape=(2, 128), dtype=int32, numpy=
array([[ 101, 3835, 3185,  102,    0,    0,    0,    0,    0,    0,    0,
           0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
           0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
           0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
           0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
           0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
           0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
           0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
           0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
           0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
           0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
           0,    0,    0,    0,    0,    0,    0],
       [ 101, 1045, 2293, 3752, 2808,  102,    0,    0,    0,    0,    0,
           0,

- The *word_id* for **CLS** it is fixed **101** and for **SEP** it is fixed **102**, between the 101 & 102 are the individual unique_id's for the word's and this could be id form the vocabulary.  
 **This is part of the preprocessing stage and in next stage is to create wordembedding and so_.on..**

In [9]:
bert_result = bert_encoder(text_preprocess)            #supply text_preprocessing
bert_result.keys()                                          # bert_result is gona be in dictionary

dict_keys(['sequence_output', 'encoder_outputs', 'default', 'pooled_output'])

- *The encoder's outputs are the **"pooled_output"** to represents each input sequence as a whole, and the **"sequence_output"** to represent each input token in context. Either of those can be used as input to further model building.*

In [10]:
bert_result['pooled_output']

<tf.Tensor: shape=(2, 768), dtype=float32, numpy=
array([[-0.82307994, -0.3302166 ,  0.05804183, ...,  0.15002297,
        -0.5304849 ,  0.8495955 ],
       [-0.85588014, -0.2829506 ,  0.00608959, ...,  0.15767564,
        -0.62548584,  0.88712823]], dtype=float32)>

- "pooled_output": a float32 Tensor of shape [batch_size, dim] with the embedding of each input sequence as a whole, derived from sequence_output in some trainable manner.

- It is embedding for the entier sentence
- We know about the 2(shape) which is the 2 sentence. What is that 768 ..?
     - 768 represents the **vector_size** (the embedding vector_size) the vector_size 768 accuratly represent the 2-statements in form of numbers, bert as generate meaningfull vector out of the statement.
     - Now we can use this vector in nlp task.

In [11]:
bert_result['sequence_output']

<tf.Tensor: shape=(2, 128, 768), dtype=float32, numpy=
array([[[-0.04972755, -0.16134058, -0.04497375, ..., -0.19434923,
          0.15166745,  0.07051349],
        [ 0.40138048, -0.71619916,  0.7960523 , ..., -0.34071296,
          0.01467817, -0.2820489 ],
        [ 0.15850015, -0.13599618, -0.3122563 , ..., -0.311211  ,
         -0.10993702, -0.5535933 ],
        ...,
        [-0.07264659, -0.22785553,  0.50889325, ..., -0.02908406,
          0.10880314,  0.17614338],
        [-0.35652003, -0.6121791 ,  0.10107114, ...,  0.15914942,
          0.01639713, -0.17355037],
        [-0.08667839, -0.28234947,  0.47228283, ...,  0.03270616,
          0.1036502 ,  0.07469998]],

       [[ 0.30540267,  0.21896993, -0.13581005, ..., -0.09990489,
          0.18915819,  0.15679148],
        [ 0.4269296 ,  0.14352971, -0.40460443, ..., -0.20114142,
          0.7126101 , -0.12644345],
        [ 1.0143212 ,  0.7115688 ,  0.28236336, ...,  0.23428604,
          0.21727912, -0.2049999 ],
        ...,

- **"sequence_output"**: a float32 Tensor of shape [batch_size, seq_length, dim] with the context-aware embedding of each token of every packed input sequence.
1. it is a individual word embedding vector
   - the size = 2, for 2 sentence for each senctence, each word 768 vector
2. What is 128 ..?
    - 128 is seq_length, for each sentence we have padding 128
    - For each of this word their is 768 size vector
    - for nice this is the vector [-0.04972755, -0.16134058, -0.04497375, ..., -0.19434923, 0.15166745,  0.07051349],
    - movie [ 0.40138048, -0.71619916,  0.7960523 , ..., -0.34071296, 0.01467817, -0.2820489 ],
    -  wait a minute, why the padding also have the number ..?
           - Because it is contextualized embedding, the vector for padding will have context of the sentence[Contextual embeddings assign a vector to each "token".]
           

In [12]:
len(bert_result['encoder_outputs'])

12

- if you the length of the encoder_outputs it is 12, the reason is 12 because we are using **BERT BASE VERSION**
     - we have 12 encoder in each layer we have 768 size embedding vector
- Encoder_outputs are output of each individual encoder which we have 12, and ech of them 12 if we look at

In [13]:
# Look at 1st one

bert_result['encoder_outputs'][0]

<tf.Tensor: shape=(2, 128, 768), dtype=float32, numpy=
array([[[ 0.1353064 ,  0.03766439, -0.06827549, ...,  0.04103095,
          0.0873357 , -0.01989367],
        [ 1.0477227 ,  1.0093771 ,  1.1757256 , ...,  0.51764846,
         -0.01262756, -0.45692933],
        [-0.06056276,  0.48112586, -0.07248963, ...,  0.6646982 ,
          0.7341418 , -0.950563  ],
        ...,
        [-0.11729388, -0.24713397,  0.6668384 , ...,  0.5240912 ,
         -0.07797258,  0.13272873],
        [-0.21674249, -0.26885363,  0.5428592 , ...,  0.55586183,
          0.00172372,  0.03732114],
        [-0.10257089, -0.19211394,  0.5575123 , ...,  0.8435401 ,
         -0.278187  , -0.02471282]],

       [[ 0.22008246,  0.056191  , -0.07639271, ...,  0.04418512,
          0.08938991,  0.04973042],
        [ 0.5641079 ,  1.1693724 , -0.21900474, ...,  0.58388084,
          0.69806415,  0.05612411],
        [ 1.3346661 ,  0.7997389 ,  0.40960526, ...,  0.6379217 ,
          0.8347206 , -0.14927903],
        ...,

- you know that (2, 128, 768), 
    1. **2** *because whe have 2 sentences*
    2. **128** *because the statement as 128 words including the padding*
    3. **768** *for each word it as 768 size embedding vector*
    
- the last layer if you see the 12th one    

In [14]:
bert_result['encoder_outputs'][11] == bert_result['sequence_output']

<tf.Tensor: shape=(2, 128, 768), dtype=bool, numpy=
array([[[ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True],
        ...,
        [ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True]],

       [[ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True],
        ...,
        [ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True]]])>

- the 12th layer, i know i have typed 11 because it start from 0 that mean the 12th layer is 11
- the 12th layer is same as your sequence output, if you compare that..they will be same

In [15]:
def get_sentence_embedding(sentence):
    preprocessed_text = bert_preprocess_model(sentence)
    return bert_encoder(preprocessed_text)['pooled_output']

In [27]:
e = get_sentence_embedding([
    
    # Fruits names
    'apple', 'banana',
    'grape', 'orange',
])

In [28]:
e

<tf.Tensor: shape=(4, 768), dtype=float32, numpy=
array([[-0.81964564, -0.29609612,  0.20951706, ...,  0.25593337,
        -0.58743   ,  0.8434557 ],
       [-0.760692  , -0.1421939 ,  0.49604598, ...,  0.42165306,
        -0.53221405,  0.80312175],
       [-0.854085  , -0.17408374,  0.23000933, ...,  0.22893815,
        -0.5581087 ,  0.851655  ],
       [-0.8363078 , -0.23830137,  0.3845352 , ...,  0.45564723,
        -0.6078616 ,  0.82788914]], dtype=float32)>

In [19]:
from sklearn.metrics.pairwise import cosine_similarity

#cosine similarity is a way to measure how similar the two vectors are to each

In [29]:
cosine_similarity([e[0]], [e[1]]) #Apple , Banana -- 96% similar  

array([[0.9659778]], dtype=float32)

>- BERT Article: By **_Jay Alammar_** explained in visual way [BERT](https://jalammar.github.io/illustrated-bert/)  

>- Bert & TensorFlow Hub link - [BERT](https://www.kaggle.com/models/tensorflow/bert/frameworks/tensorFlow2/variations/en-uncased-l-12-h-768-a-12), [TF_HUB](https://www.tensorflow.org/hub)  

>- To learn about this bert en-uncased-Preprocessor. click - [PREPROCEESSOR](https://www.kaggle.com/models/tensorflow/bert/frameworks/TensorFlow2/variations/en-uncased-preprocess/versions/3)

>- To learn about bert Model APIs - [MODEL API](https://www.tensorflow.org/hub/common_saved_model_apis/text#transformer-encoders)